[Feat] Introduce NodeType::BootstrapClient

ljedrz opened the pull request:
This PR introduces a new `NodeType::BootstrapClient`, which is a light node that can serve peer lists to all kinds of nodes. It also adjusts the `OnConnect` behavior of other node types to immediately request a list of peers when connecting to one.

I tested these changes by starting a `--dev N --bootstrap-client` node, and connecting to it in both non-validator and validator handshake mode from other `--dev` nodes that had it specified as one of their `TEST_BOOTSTRAP_PEERS`; I also hardcoded the `mainnet` network once to ensure that the latest validator committee is downloaded and deserialized correctly.

Open points / suggestions:
- the `bootstrap_peers` function could now be removed in favor of introducing candidate peers at node's startup (`CandidatePeer` could have an `Option<NodeType>` member, which could also be useful when refreshing peers etc)
- the `Event::ChallengeRequest` object is lacking a `NodeType` member, making it both incompatible with the `Message::ChallengeRequest` one and forcing us to determine the node type based on our list of bootstrap nodes

Stress-tests also succeeded after several iterations.

Closes https://github.com/ProvableHQ/snarkOS/issues/3868.

ljedrz committed:
feat: introduce a BootstrapClient node type

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

ljedrz committed:
refactor: add PeerPoolHandling::get_best_connected_peers

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

vicsn commented on node/router/messages/src/helpers/node_type.rs:
```diff
@@ -77,6 +81,7 @@ impl FromBytes for NodeType {
             1 => Ok(Self::Prover),
             2 => Ok(Self::Validator),
+            3 => Ok(Self::BootstrapClient),
             x => Err(error(format!("Invalid node type: expected 0, 1, or 2, got {x}."))),
```
```suggestion
            x => Err(error(format!("Invalid node type: expected 0..3, got {x}."))),
```

vicsn commented on node/router/src/lib.rs:
```diff
@@ -669,7 +694,7 @@ impl<N: Network> Router<N> {
         let lowest_accepted_message_version = match self.node_type {
             // Provers should always use the latest version.
-            NodeType::Prover => Message::<N>::latest_message_version(),
+            NodeType::Prover | NodeType::BootstrapClient => Message::<N>::latest_message_version(),
```
I think it's better to be a bit more lenient and to let the bootstrap client accept messages from lower versions based on migration height (i.e. move to the match arm below.

ljedrz replied:
we don't know the current block height, though

vicsn replied:
Good point, can you document that in the comment?

vicsn commented on node/src/bootstrap_client/handshake.rs:
```diff
@@ -0,0 +1,321 @@
+            debug!("Received a connection request from '{peer_addr}'");
+            None
+        } else {
+            unreachable!("The boostrapper clients don't initiate connections");
```
Also not if peers are passed via the --peers or --validators flag? Maybe to prevent this footgun we should disallow setting peers or validators in the start command if --bootstrap-client is set.

vicsn commented on node/src/bootstrap_client/network.rs:
```diff
@@ -0,0 +1,173 @@
+            }
+            MessageOrEvent::Event(Event::ValidatorsRequest(_)) => {
+                debug!("Received a ValidatorsRequest from '{listener_addr}'");
+                let current_committee = match self.get_or_update_committee().await {
```
In dev mode, i.e. in ephemeral devnets, we won't be able to get the committee from the block explorer.

Perhaps we can disable this filtering if dev mode is enabled

vicsn commented on node/src/bootstrap_client/network.rs:
```diff
@@ -0,0 +1,173 @@
+        match message {
+            MessageOrEvent::Message(Message::PeerRequest(_)) => {
+                debug!("Received a PeerRequest from '{listener_addr}'");
+                let peers = self.get_best_connected_peers(Some(u8::MAX as usize));
```
Can we re-use MAX_PEERS_TO_SEND?

vicsn commented on node/src/bootstrap_client/network.rs:
```diff
@@ -0,0 +1,173 @@
+                let _ = self.unicast(peer_addr, msg)?.await;
+
+                debug!("Disconnecting from '{listener_addr}' - peers provided");
+                self.tcp().disconnect(peer_addr).await;
```
Savage

vicsn commented on node/src/bootstrap_client/network.rs:
```diff
@@ -0,0 +1,173 @@
+                self.tcp().disconnect(peer_addr).await;
+            }
+            _ => {
+                trace!("Ignoring an unhandled message from {listener_addr}");
```
Maybe still nice to print the message id

ljedrz committed:
tweak: apply review comments

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

ljedrz committed:
chore: adjust the lockfile

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

ljedrz commented:
@vicsn I think I've addressed all your comments; one of them also helped me squash a bug (wrong identification of validator-mode handshakes, which is now fixed with `validator_mode`), which is nice.

vicsn commented on node/src/bootstrap_client/network.rs:
```diff
@@ -129,20 +133,25 @@ impl<N: Network> Reading for BootstrapClient<N> {
+                        committee.contains(&peer.aleo_addr)
+                    } else {
+                        // Otherwise, filter out known peers connected in validator mode (via Gateway).
+                        self.resolve_to_aleo_addr(peer.listener_addr).is_some()
```
Does this mean we are only sending validators which we are **currently** connected to? Given the bootstrap client's rapid disconnecting, that would be very little. We should just return all of the validators we know.

Let me know if I misunderstand.

ljedrz replied:
No, you're spot on: https://github.com/ProvableHQ/snarkOS/pull/3920/commits/c7e37d3c72735b3a01e70ee3947fd4cf807b85f9.

ljedrz committed:
tweak: apply review comments

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

ljedrz committed:
feat: make validators periodically connect to bootstrap clients

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

ljedrz committed:
fix: correct a bootstrapper handshake log, don't reject connected bootstrappers in Gateway

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

ljedrz committed:
fix: reject multiple concurrent peer list insertions

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

ljedrz committed:
fix: properly fix a deadlock in PeerPoolHandling::insert_candidate_peers

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

ljedrz committed:
fix: accept Events from bootstrappers, don't put their Aleo addrs in the resolver

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

ljedrz committed:
tweak: log any send errors, add a small delay

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

ljedrz committed:
tweak: remove the debugging delay

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

ljedrz committed:
tweak: don't special-case validators when shuffling peers in the Router

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

vicsn commented on node/router/src/heartbeat.rs:
```diff
@@ -119,8 +120,6 @@ pub trait Heartbeat<N: Network>: Outbound<N> {
     ///     - Validators are considered higher priority than provers or clients.
```
This comment can be removed

ljedrz committed:
tweak: apply review comments

Signed-off-by: ljedrz <ljedrz@users.noreply.github.com>

vicsn reviewed (approved):
LGTM!
